Refresh token impl (#2212)
wasp-lang/wasp: web/docs/auth/auth-hooks.md

The `onAfterSignup` hook can be useful if you want to send the user a welcome email or perform some other action after the user signs up like syncing the user with a third-party service.

> Since the `onAfterSignup` hook receives the OAuth tokens, it can also be used to store the OAuth access token and/or [refresh token](#refreshing-the-oauth-access-token) in your database.

Reviewer: Passive voice. Please fix in other places as well.

Author: I keep writing using passive voice!

  console.log('number of users after', count)
  console.log('user object', user)

>   // If this is an OAuth signup, we have the OAuth tokens and uniqueRequestId

Reviewer: Address the user directly plus "the" before `uniqueRequestId`. Same applies below.

The `onAfterLogin` hook can be useful if you want to perform some action after the user logs in, like syncing the user with a third-party service.

> Since the `onAfterLogin` hook receives the OAuth tokens, you can also use it to update the OAuth access token for the user in your database. You can also use it to [refresh the OAuth access token](#refreshing-the-oauth-access-token) if the provider supports it.

Reviewer: Redundant "also" in the first sentence I think (makes the second sentence seeem repetitive).

Author: I already have this version locally, this is some weird Github diff.

The `onAfterLogin` hook can be useful if you want to perform some action after the user logs in, like syncing the user with a third-party service.

> Since the `onAfterLogin` hook receives the OAuth tokens, it can also be used to update the OAuth access token for the user in your database. It can also be used to [refresh the OAuth access token](#refreshing-the-oauth-access-token) if the provider supports it.

Reviewer: Passive voice.

Here's an example of how you can refresh the access token for Google OAuth:

<Tabs groupId="js-ts">

<TabItem value="js" label="JavaScript">

```js title="src/auth/hooks.js"
import { google } from 'wasp/server/oauth'

export const onAfterLogin = async ({ oauth }) => {
  if (oauth.provider === 'google' && oauth.tokens.refreshToken !== null) {
    const newTokens = await google.oAuthClient.refreshAccessToken(
      oauth.tokens.refreshToken
    )
    log('new tokens', newTokens)
  }
}
```

</TabItem>

<TabItem value="ts" label="TypeScript">

```ts title="src/auth/hooks.ts"
import type { OnAfterLoginHook } from 'wasp/server/auth'
import { google } from 'wasp/server/oauth'

export const onAfterLogin: OnAfterLoginHook = async ({ oauth }) => {
  if (oauth.provider === 'google' && oauth.tokens.refreshToken !== null) {
    const newTokens = await google.oAuthClient.refreshAccessToken(
      oauth.tokens.refreshToken
    )
    log('new tokens', newTokens)
  }
}
```

</TabItem>
</Tabs>

> Google exposes the `accessTokenExpiresAt` field in the `oauth.tokens` object. You can use this field to determine when will the access token expire.

Reviewer: ```suggestion
Google exposes the `accessTokenExpiresAt` field in the `oauth.tokens` object. You can use this field to determine when the access token expires.
```
Based on preferring present tense, and it's shorter.

Read more about the data the `onAfterLogin` hook receives in the [API Reference](#the-onafterlogin-hook).

### Refreshing the OAuth access token

> If you are using OAuth, you can use the `onAfterSignup` or `onAfterLogin` hooks to refresh the OAuth access token if the provider supports it.

Reviewer: Which of these hooks should they use for this and why?

Can we recommend/explain it?

Author: I've rewritten this section to be more clear, thanks for the nudge 👍

> You'll need to import the OAuth client from the `wasp/server/oauth` module for the provider you are using. The OAuth client has a `refreshAccessToken` method that you can use to refresh the access token.

Reviewer: ```suggestion
Import the OAuth client for the desired provider from `wasp/server/oauth` and use its `refreshAccessToken` method to refresh the access token.
```
Suggestions based on:
 - Addressing the user directly
 - Prefer present tense
 - Prefer shorter and simpler sentences

> Here is an example of how you can refresh the Google OAuth access token:

Reviewer: ```suggestion
Here's an example of how you can refresh the access token for Google OAuth:
```
Subjective, but since we're talking about "refreshing the access token" all the time, I recommend keeping that phrase together, and appending the specific provider at the end.

<Tabs groupId="js-ts">

<TabItem value="js" label="JavaScript">

```js title="src/auth/hooks.js"
import { google } from 'wasp/server/oauth'

export const onAfterLogin = async ({ oauth }) => {
  if (oauth.provider === 'google' && oauth.tokens.refreshToken !== null) {
    const newTokens = await google.oAuthClient.refreshAccessToken(
      oauth.tokens.refreshToken
    )
    log('new tokens', newTokens)
  }
}
```
</TabItem>

<TabItem value="ts" label="TypeScript">

```ts title="src/auth/hooks.ts"
import type { OnAfterLoginHook } from 'wasp/server/auth'
import { google } from 'wasp/server/oauth'

export const onAfterLogin: OnAfterLoginHook = async ({ oauth }) => {
  if (oauth.provider === 'google' && oauth.tokens.refreshToken !== null) {
    const newTokens = await google.oAuthClient.refreshAccessToken(
      oauth.tokens.refreshToken
    )
    log('new tokens', newTokens)
  }
}
```
</TabItem>
</Tabs>

> Google provides you with the `accessTokenExpiresAt` field in the `oauth.tokens` object. You can use this field to determine when the access token expires and refresh it before it expires.

Reviewer: ```suggestion
Google exposes the `accessTokenExpiresAt` field in the `oauth.tokens` object. You can use this field to determine when the access token expires.
```
Subjective, but I like it because it's shorter.

I removed the "and refresh it before it expires." part because it suggests that they can use the field to refresh the token, which I don't think is right.

> You can of course perform the token refresh in an [Wasp Job](../advanced/jobs.md) if you want to refresh the token periodically.

Reviewer: ```suggestion
If you want to refresh the token periodically, use a [Wasp Job](../advanced/jobs.md).
```
Suggestion based on:
 - Start with a problem, follow with a solution.
 - Shorter sentences.

It has the following fields:

- `providerName: string`

>   The OAuth provider name (e.g. `'google'`, `'github'`) that the user authenticated with.

Reviewer: ```suggestion
  The name of the OAuth provider the user authenticated with (e.g. `'google'`, `'github'`).
```
This formulation is less confusing IMO. The original soulds like the user authenticated with an "OAuth provider name."
Plus, the examples are in the middle of the sentence, breaking the flow.

- `tokens: Tokens`

  You can use the OAuth tokens to make requests to the provider's API on the user's behalf.

>   Depending on the OAuth provider, the `tokens` object might have different fields. For example, for Google, it has the `accessToken`, `refreshToken`, `idToken`, and `accessTokenExpiresAt` fields.

Reviewer: ```suggestion
  Depending on the OAuth provider, the `tokens` object might have different fields. For example, Google has the fields `accessToken`, `refreshToken`, `idToken`, and `accessTokenExpiresAt`.
```
Suggestions based on:
 - Fewer commas make a simpler sentence.
 - Listing stuff at the end of the sentence improves the flow.

  <ShowForTs>

>   You can narrow down the `oauth.tokens` object type to the specific OAuth provider type to access the provider-specific fields:

Reviewer: ```suggestion
To access the provider-specific fields, you must first narrow down the `oauth.tokens` object type to he specific OAuth provider type.
```
Suggestions based on:
- Start with a problem, follow with a solution.
- can -> must (because they must :))
